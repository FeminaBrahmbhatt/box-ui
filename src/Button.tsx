import React, { ReactElement, cloneElement, isValidElement } from "react";
import {
  ButtonProps,
  ButtonVariant,
  ButtonSize,
  ButtonState,
} from "./@interfaces";
import classNames from "classnames";

const buttonVariantClasses: Record<
  ButtonVariant,
  Record<ButtonState, string>
> = {
  primary: {
    default: "btn-primary",
    hover: "btn-primary-hover",
    focus: "btn-primary-focus shadow-grayDark",
    disabled: "btn-primary-disabled",
  },
  secondary: {
    default: "btn-secondary",
    hover: "btn-secondary-hover",
    focus: "btn-secondary-focus shadow-grayDark",
    disabled: "btn-secondary-disabled",
  },
  secondaryGray: {
    default: "btn-secondaryGray",
    hover: "btn-secondaryGray-hover",
    focus: "btn-secondaryGray-focus shadow-grayDark",
    disabled: "btn-secondaryGray-disabled",
  },
  tertiary: {
    default: "btn-tertiary",
    hover: "btn-tertiary-hover",
    focus: "btn-tertiary-focus shadow-grayDark",
    disabled: "btn-tertiary-disabled",
  },
  tertiaryGray: {
    default: "btn-tertiaryGray",
    hover: "btn-tertiaryGray-hover",
    focus: "btn-tertiaryGray-focus shadow-grayDark",
    disabled: "btn-tertiaryGray-disabled",
  },
};

const buttonSizeClasses: Record<ButtonSize, string> = {
  sm: "btn-sm",
  md: "btn-md",
  lg: "btn-lg",
  xl: "btn-xl",
  "2xl": "btn-2xl",
};

const buttonIconSizeClasses: Record<ButtonSize, string> = {
  sm: "btn-icon-sm",
  md: "btn-icon-md",
  lg: "btn-icon-lg",
  xl: "btn-icon-xl",
  "2xl": "btn-icon-2xl",
};

export const Button = ({
  children,
  className,
  variant = "primary",
  size = "md",
  disabled,
  LeadingIcon,
  TrailingIcon,
  IconOnly,
  ...buttonProps
}: ButtonProps) => {
  const buttonVariantClassName = buttonVariantClasses[variant];
  const buttonIconSizeClassName = buttonIconSizeClasses[size];

  const setIconSize = (icon: ReactElement, iconSize: number) => {
    if (isValidElement<{ size?: number }>(icon) && "size" in icon.props) {
      return cloneElement(icon, { size: 20 });
    }
    return icon;
  };

  const setClassName = (icon: ReactElement, additionalClass: string) => {
    return isValidElement<{ className?: string }>(icon)
      ? cloneElement(icon, {
          className: classNames(icon.props.className, additionalClass),
        })
      : icon;
  };

  return (
    <button
      type="button"
      className={classNames("btn-base", className, {
        [buttonSizeClasses[size]]: !IconOnly,
        [classNames(buttonIconSizeClassName, "justify-center")]: IconOnly,
        [classNames(buttonVariantClassName.default,
        buttonVariantClassName.hover,
        buttonVariantClassName.focus)]: !disabled,
        [classNames(buttonVariantClassName.disabled, "cursor-not-allowed")]: disabled,
      })}
      {...buttonProps}
    >
      {LeadingIcon && (
        <span>
          {setClassName(
            LeadingIcon,
            classNames({
              "mr-2": size !== "2xl",
              "mr-3": size === "2xl",
            })
          )}
        </span>
      )}
      {children}
      {IconOnly && (
        <span>{setIconSize(IconOnly, size === "2xl" ? 24 : 20)}</span>
      )}
      {TrailingIcon && (
        <span>
          {setClassName(
            TrailingIcon,
            classNames({
              "ml-2": size !== "2xl",
              "ml-3": size === "2xl",
            })
          )}
        </span>
      )}
    </button>
  );
};
