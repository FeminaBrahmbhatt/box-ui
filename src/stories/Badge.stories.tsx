import React from 'react';
import {BadgeProps} from '../@interfaces';
import { Badge } from '../Badge';
import {images} from '../data/images'
import {Meta, StoryObj} from '@storybook/react';
import { StoryLayout } from '../StoryLayout';
import { FiArrowRight, FiStar } from 'react-icons/fi';
import { IconWrapper } from '../utils/IconWrapper';

export default {
    title: 'Example/Badge',
    component: Badge,
    parameters: {
        layout: 'centered',
    },
    tags: ['autodocs'],
} as Meta;
interface Props extends BadgeProps {
    darkMode: boolean;
}

export const Default:StoryObj<Props> = {
    args: {
        darkMode: false,
        variant: "success",
        size: "md",
        children:<span>Label</span>
    }
}

export const LeadingIconBadge:StoryObj<Props>  = {
    args: {
        darkMode: false,
        variant: "success",
        size: "md",
        LeadingIcon: <IconWrapper Icon={FiStar} data-testid="leadingIcon"/>
    },
    render: (args) => {
        return (
            <StoryLayout {...args} className="inline-flex flex-col space-y-2">
                <Badge {...args}>
                    Label
                </Badge>
            </StoryLayout>
        )
    }
}

export const ImageIconBadge:StoryObj<Props>  = {
    args: {
        darkMode: false,
        variant: "gray",
        size: "md",
        LeadingIcon: <img src={images.NL} alt="nl" className="w-4 h-4" data-testid="iconOnly" />
    },
    render: (args) => {
        return (
            <StoryLayout {...args} className="inline-flex flex-col space-y-2">
                <Badge {...args}>
                    Newzeland
                </Badge>
            </StoryLayout>
        )
    }
}

export const TrailingIconBadge:StoryObj<Props>  = {
    args: {
        darkMode: false,
        variant: "primary",
        size: "md",
        TrailingIcon: <IconWrapper Icon={FiArrowRight} data-testid="trailingIcon"/>
    },
    render: (args) => {
        return (
            <StoryLayout {...args} className="inline-flex flex-col space-y-2">
                <Badge {...args}>
                    Label
                </Badge>
            </StoryLayout>
        )
    }
}

