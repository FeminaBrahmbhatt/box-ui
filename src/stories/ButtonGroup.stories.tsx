import React, { useState } from "react";
import { Meta, StoryObj } from "@storybook/react";
import { ButtonGroupProps, ButtonGroup } from "../ButtonGroup";
import { StoryLayout } from "../StoryLayout";
import { options1, options2, options3 } from "../data/options";

export default {
  title: "Example/ButtonGroup",
  component: ButtonGroup,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
} as Meta;
interface Props extends ButtonGroupProps {
  darkMode: boolean;
}

const LabelButtonGroupComponent = (args: Props) => {
  const [activeItem1, setActiveItem1] = useState<string>(options1[1].value);

  return (
    <StoryLayout {...args}>
      <ButtonGroup
        {...args}
        options={options1}
        activeOption={activeItem1}
        setActiveOption={setActiveItem1}
      />
    </StoryLayout>
  );
};

export const LabelButtonGroup: StoryObj<Props> = {
  args: {
    darkMode: false,
  },
  render: (args) =>  <LabelButtonGroupComponent {...args} />,
};

const IconButtonGroupComponent = (args: Props) => {
  const [activeView, setViewOption] = useState<string>("list");

  return (
    <StoryLayout {...args}>
      <ButtonGroup
        {...args}
        options={options3}
        activeOption={activeView}
        setActiveOption={setViewOption}
      />
    </StoryLayout>
  );
};

export const IconButtonGroup: StoryObj<Props> = {
  args: {
    darkMode: false,
  },
  render: (args) => <IconButtonGroupComponent {...args} />,
};

const LabelIconButtonGroupComponent = (args: Props) => {
  const [activeItem2, setActiveItem2] = useState<string>(options2[1].value);

  return (
    <StoryLayout {...args}>
      <ButtonGroup
        {...args}
        options={options2}
        activeOption={activeItem2}
        setActiveOption={setActiveItem2}
      />
    </StoryLayout>
  );
};

export const LabelIconButtonGroup: StoryObj<Props> = {
  args: {
    darkMode: false,
  },
  render: (args) => <LabelIconButtonGroupComponent {...args} />,
};
