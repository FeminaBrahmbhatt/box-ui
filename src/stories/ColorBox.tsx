import React from "react";
import { IColor } from "./@interfaces";

export const ColorBox = ({ code, hex, bgClass }: IColor) => (
<div
    key={bgClass}
    className="flex flex-col rounded-lg shadow-lg w-26 h-39 dark:bg-white">
      <div className={`h-20 ${bgClass} rounded-t-lg`} />
      <div className="mx-3 my-3">
          {code}
          <br/>
          {hex}
      </div>
    </div>
);