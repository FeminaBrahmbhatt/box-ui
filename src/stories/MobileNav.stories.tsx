import React, { useState } from "react";
import { Meta, StoryObj } from "@storybook/react";
import { MobileNavProps } from "../@interfaces";
import { SideNav } from "../Nav/SideNav";
import { StoryLayout } from "../StoryLayout";
import { navItemsTop, navItemsBottom } from "../data/navItems";
import { MobileNavbar } from "../Nav/MobileNav";
import { FiX } from "react-icons/fi"; // Directly import FiX
import type { IconBaseProps } from "react-icons";

const CloseIcon = FiX as React.FC<IconBaseProps>;
// Wrapper to fix TS JSX issues
 
export default {
  title: "Example/MobileNav",
  component: MobileNavbar,
} as Meta;

interface Props extends MobileNavProps {
  darkMode: boolean;
}

const MobileNavWithState = (args: Props) => {
  const [open, setOpen] = useState<boolean>(false);

  const handleToggle = () => {
    setOpen(!open);
  };

  return (
    <StoryLayout {...args}>
      <MobileNavbar open={open} toggleOpen={handleToggle} />

      {open && (
        <div
          data-testid="backdrop"
          className="fixed top-0 left-0 z-40 w-screen h-screen bg-gray-500 cursor-pointer bg-opacity-80"
          onClick={() => setOpen(false)}
        >
           <CloseIcon
            data-testid="close-icon"
            size={40}
            className="fixed text-white top-5 right-9"
            role="button"
            aria-label="Close sidebar"
          />
        </div>
      )}

      {open && (
        <SideNav
          className="relative z-50 h-screen -mt-20"
          navItemsTop={navItemsTop}
          navItemsBottom={navItemsBottom}
          open={true}
          setOpen={handleToggle}
          username="Veronica Woods"
          email="veronica@example.com"
        />
      )}
    </StoryLayout>
  );
};

export const Default: StoryObj<Props> = {
  args: {
    darkMode: false,
  },
  render: (args) => <MobileNavWithState {...args} />,
};
