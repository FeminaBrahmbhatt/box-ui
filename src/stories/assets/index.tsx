
export type CountryISO = "BE" | "DE" | "FR" | "IT" | "NL" | "RU" | "US";

type image = CountryISO;

// Development images
const devImages: Record<string, image> = {
  NL: require("./nl.svg")
};

// Production images (CDN)
// const prodImages: Record<string, image> = {
//   NL: "https://res.cloudinary.com/tailwindcss/image/upload/v1635279277/nl_tpy2ab.svg",
// };

export const assets = devImages;