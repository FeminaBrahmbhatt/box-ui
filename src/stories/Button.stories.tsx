import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import { Button } from "../Button";
import { ButtonProps } from "../@interfaces";
import { FiArrowLeft, FiArrowRight, FiStar } from "react-icons/fi";
import { StoryLayout } from "../StoryLayout";
import { IconWrapper } from "../utils/IconWrapper";

export default{
  title: "Example/Button",
  component: Button,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
} as Meta;
interface Props extends ButtonProps {
  darkMode: boolean;
}

type Story = StoryObj<Props>;

export const Default: Story = {
  args: {
    darkMode: false,
    children: "Button",
    size: 'md',
    variant: 'primary',
    disabled: false,
    LeadingIcon: undefined,
    TrailingIcon: undefined,
    IconOnly: undefined,
    className: ''
  },
  render: (args) => {
    return (
      <StoryLayout {...args} className="flex inline-flex flex-row space-x-2">
        <Button {...args}>Primary</Button>
      </StoryLayout>
    );
  },
};

export const Variants: Story = {
  args: {
    darkMode: false,
    variant: "primary",
  },
  render: (args) => {
    return (
      <StoryLayout {...args} className="flex inline-flex flex-row space-x-2">
        <Button variant="primary">Primary</Button>
        <Button variant="secondary">Secondary</Button>
        <Button variant="secondaryGray">Secondary Gray</Button>
        <Button variant="tertiary">Tertiary</Button>
        <Button variant="tertiaryGray">Tertiary Gray</Button>
      </StoryLayout>
    );
  },
};

export const Sizes: Story = {
  args: {
    darkMode: false,
    variant: "primary",
    size: "md",
  },
  render: (args) => {
    return (
      <StoryLayout {...args} className="flex inline-flex flex-row space-x-2">
        <Button size="sm">sm</Button>
        <Button size="md">md</Button>
        <Button size="lg">lg</Button>
        <Button size="xl">xl</Button>
        <Button size="2xl">2xl</Button>
      </StoryLayout>
    );
  },
};

export const TrailingIconButton: Story = {
  args: {
    darkMode: false,
    children: "Button",
    TrailingIcon: <IconWrapper Icon={FiArrowRight} data-testid="trailingIcon"/>,
  },  
  render: (args) => (
    <StoryLayout {...args}>
      <Button {...args} />
    </StoryLayout>
  ),
};

export const LeadingIconButton: Story = {
  args: {
    darkMode: false,
    children: "Button",
    LeadingIcon: <IconWrapper Icon={FiArrowLeft} data-testid="LeadingIcon"/>,
  },
  render: (args) => (
    <StoryLayout {...args}>
      <Button {...args} />
    </StoryLayout>
  ),
};

export const IconOnlyButton: Story = {
  args: {
    darkMode: false,
    IconOnly: <IconWrapper Icon={FiStar} data-testid="iconOnly"/>,
  },
  render: (args) => (
    <StoryLayout {...args}>
      <Button {...args} />
    </StoryLayout>
  ),
};
