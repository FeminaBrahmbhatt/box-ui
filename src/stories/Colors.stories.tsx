import React from "react";
import { Meta, StoryObj } from "@storybook/react";
import { ColorBox } from "../ColorBox";
import { colors } from "../data/colors";
import { StoryLayout } from "../StoryLayout";

export default {
  title: "Colors",
  parameters: {
    controls: { expanded: true },
  },
} as Meta;

interface Props {
  darkMode: boolean;
}

export const Default: StoryObj<Props> = {
  args: {
    darkMode: false,
  },
  render: (args) => (
    <StoryLayout {...args} className="grid grid-cols-2 gap-4 md:grid-cols-4 lg:grid-cols-6 xl:grid-cols-11">
      {colors.map((color) => (
        <ColorBox key={color.bgClass} code={color.code} hex={color.hex} bgClass={color.bgClass}/>
      ))}
    </StoryLayout>
  ),
};
