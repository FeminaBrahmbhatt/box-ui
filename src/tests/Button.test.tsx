
import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/react';
import * as stories from '../stories/Button.stories';

const {
    Default, Variants, Sizes, LeadingIconButton, TrailingIconButton, IconOnlyButton
  } = composeStories(stories);

describe('ðŸ“˜ Button Storybook Stories', () => {
    test('renders Default button', () => {
      const Story = Default;
      const { container } = render(
        Story(Story.args ?? {})
      );
      expect(screen.getByText('Primary')).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  
    test('renders all button variants', () => {
      const Story = Variants;
      const { container } = render(
        Story(Story.args ?? {})
      );
      ['Primary', 'Secondary', 'Secondary Gray', 'Tertiary', 'Tertiary Gray'].forEach(label => {
        expect(screen.getByText(label)).toBeInTheDocument();
      });
      expect(container).toMatchSnapshot();
    });
  
    test('renders all button sizes', () => {
      const Story = Sizes;
      const { container } = render(
        Story(Story.args ?? {})
      );
      ['sm', 'md', 'lg', 'xl', '2xl'].forEach(label => {
        expect(screen.getByText(label)).toBeInTheDocument();
      });
      expect(container).toMatchSnapshot();
    });
  
    test('renders button with trailing icon', () => {
      const Story = TrailingIconButton;
      const { container } = render(
        Story(Story.args ?? {})
      );
      expect(screen.getByText('Button')).toBeInTheDocument();
      expect(container.querySelector('svg')).toBeInTheDocument(); // FiArrowRight
    });
  
    test('renders button with leading icon', () => {
      const Story = LeadingIconButton;
      const { container } = render(
        Story(Story.args ?? {})
      );
      expect(screen.getByText('Button')).toBeInTheDocument();
      expect(container.querySelector('svg')).toBeInTheDocument(); // FiArrowLeft
    });
  
    test('renders icon-only button', () => {
      const Story = IconOnlyButton;
      const { container } = render(
        Story(Story.args ?? {})
      );
      expect(container.querySelector('svg')).toBeInTheDocument(); // FiStar
    });
  });