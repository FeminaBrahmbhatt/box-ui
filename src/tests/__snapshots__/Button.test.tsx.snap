// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ“˜ Button Storybook Stories renders Default button 1`] = `
<div>
  <div
    class="-m-4"
  >
    <div
      class="flex inline-flex flex-row space-x-2 p-4"
    >
      <button
        class="btn-base btn-md btn-primary btn-primary-hover btn-primary-focus shadow-grayDark"
        type="button"
      >
        Primary
      </button>
    </div>
  </div>
</div>
`;

exports[`ðŸ“˜ Button Storybook Stories renders all button sizes 1`] = `
<div>
  <div
    class="-m-4"
  >
    <div
      class="flex inline-flex flex-row space-x-2 p-4"
    >
      <button
        class="btn-base btn-sm btn-primary btn-primary-hover btn-primary-focus shadow-grayDark"
        type="button"
      >
        sm
      </button>
      <button
        class="btn-base btn-md btn-primary btn-primary-hover btn-primary-focus shadow-grayDark"
        type="button"
      >
        md
      </button>
      <button
        class="btn-base btn-lg btn-primary btn-primary-hover btn-primary-focus shadow-grayDark"
        type="button"
      >
        lg
      </button>
      <button
        class="btn-base btn-xl btn-primary btn-primary-hover btn-primary-focus shadow-grayDark"
        type="button"
      >
        xl
      </button>
      <button
        class="btn-base btn-2xl btn-primary btn-primary-hover btn-primary-focus shadow-grayDark"
        type="button"
      >
        2xl
      </button>
    </div>
  </div>
</div>
`;

exports[`ðŸ“˜ Button Storybook Stories renders all button variants 1`] = `
<div>
  <div
    class="-m-4"
  >
    <div
      class="flex inline-flex flex-row space-x-2 p-4"
    >
      <button
        class="btn-base btn-md btn-primary btn-primary-hover btn-primary-focus shadow-grayDark"
        type="button"
      >
        Primary
      </button>
      <button
        class="btn-base btn-md btn-secondary btn-secondary-hover btn-secondary-focus shadow-grayDark"
        type="button"
      >
        Secondary
      </button>
      <button
        class="btn-base btn-md btn-secondaryGray btn-secondaryGray-hover btn-secondaryGray-focus shadow-grayDark"
        type="button"
      >
        Secondary Gray
      </button>
      <button
        class="btn-base btn-md btn-tertiary btn-tertiary-hover btn-tertiary-focus shadow-grayDark"
        type="button"
      >
        Tertiary
      </button>
      <button
        class="btn-base btn-md btn-tertiaryGray btn-tertiaryGray-hover btn-tertiaryGray-focus shadow-grayDark"
        type="button"
      >
        Tertiary Gray
      </button>
    </div>
  </div>
</div>
`;
